README.md
# Status page service

[![CircleCI](https://circleci.com/gh/contartec-team/status-page-service.svg?style=shield&circle-token=1842b6b974b8c6d61af2aa365d9242da5b2c0b96)](https://circleci.com/gh/contartec-team/status-page-service)
[![Maintainability](https://api.codeclimate.com/v1/badges/51b1b80345d08d73ee46/maintainability)](https://codeclimate.com/github/contartec-team/status-page-service/maintainability)
[![deepcode](https://www.deepcode.ai/api/gh/badge?key=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwbGF0Zm9ybTEiOiJnaCIsIm93bmVyMSI6ImNvbnRhcnRlYy10ZWFtIiwicmVwbzEiOiJzdGF0dXMtcGFnZS1zZXJ2aWNlIiwiaW5jbHVkZUxpbnQiOmZhbHNlLCJhdXRob3JJZCI6MTczMjgsImlhdCI6MTYxMTY4OTQ1NX0.qsGzEjgXvjHQPVuSrki-Y_H9bUv60GfxmkbsOf57hSI)](https://www.deepcode.ai/app/gh/contartec-team/status-page-service/_/dashboard?utm_content=gh%2Fcontartec-team%2Fstatus-page-service)
[![codecov](https://codecov.io/gh/contartec-team/status-page-service/branch/master/graph/badge.svg?token=b31TRIAM3z)](https://codecov.io/gh/contartec-team/status-page-service)

Services for `Status Page`

## Install

`npm i @contartec-team/status-page-service`

## Docs

https://contartec-team.github.io/status-page-service/index

CHANGELOG.md
## 0.2.1 (2021-08-24)


### Bug Fixes

* [`StatusService`] `.updateStatus` não atualiza `StatusPage`

## 0.2.0 (2021-01-27)


### Features

* [`StatusService`] Add `headers` no `.getAPIResponse`

## 0.1.0 (2021-01-26)


### Features

* Release inicial

## 0.3.0 (2021-01-22)


### Features

* Add `GenerickMock.addList`


tests/mocha.init.js
'use strict'

require('dotenv').config()

const chai = require('chai')

chai
  .use(require('chai-things'))
  .use(require('chai-as-promised'))
  .use(require('chai-shallow-deep-equal'))
  .use(require('sinon-chai'))

global.expect = chai.expect

tests/services/StatusService.test.js
'use strict'

const rewire = require('rewire')
const StatusService = rewire('../../lib/services/StatusService')

const SpyMock = require('@contartec-team/spy-mock/lib/SpyMock')

describe('StatusService', () => {
  describe('.DEFAULT_AXIOS_PARAMS', () => {
    context('when env vars are valid', () => {
      let axiosParams = {}
      let params

      before(() => {
        axiosParams = {
          url     : process.env.API_URL,
          method  : process.env.API_METHOD,
          headers : JSON.parse(process.env.API_HEADERS),
          data    : JSON.parse(process.env.API_POST_BODY),
          params  : JSON.parse(process.env.API_QUERY_STRING)
        }

        params = StatusService.DEFAULT_AXIOS_PARAMS
      })

      it('should return `axios` params from env vars', () => {
        expect(params).to.eql(axiosParams)
      })
    })

    context('when `url` is `null`', () => {
      const url = process.env.API_URL

      let axiosParams = {}
      let params

      before(() => {
        process.env.API_URL = ''

        axiosParams = {
          url     : '',
          method  : process.env.API_METHOD,
          headers : JSON.parse(process.env.API_HEADERS),
          data    : JSON.parse(process.env.API_POST_BODY),
          params  : JSON.parse(process.env.API_QUERY_STRING)
        }

        params = StatusService.DEFAULT_AXIOS_PARAMS
      })

      after(() => process.env.API_URL = url)

      it('should return `axios` params from env vars', () => {
        expect(params).to.eql(axiosParams)
      })
    })

    context('when `API_HEADERS` is `null`', () => {
      const headers = process.env.API_HEADERS

      let axiosParams = {}
      let params

      before(() => {
        process.env.API_HEADERS = ''

        axiosParams = {
          url     : process.env.API_URL,
          method  : process.env.API_METHOD,
          headers : null,
          data    : JSON.parse(process.env.API_POST_BODY),
          params  : JSON.parse(process.env.API_QUERY_STRING)
        }

        params = StatusService.DEFAULT_AXIOS_PARAMS
      })

      after(() => process.env.API_HEADERS = headers)

      it('should return `axios` params from env vars', () => {
        expect(params).to.eql(axiosParams)
      })
    })

    context('when `API_POST_BODY` is `null`', () => {
      const body = process.env.API_POST_BODY

      let axiosParams = {}
      let params

      before(async () => {
        process.env.API_POST_BODY = ''

        axiosParams = {
          url     : process.env.API_URL,
          method  : process.env.API_METHOD,
          headers : JSON.parse(process.env.API_HEADERS),
          data    : null,
          params  : JSON.parse(process.env.API_QUERY_STRING)
        }

        params = StatusService.DEFAULT_AXIOS_PARAMS
      })

      after(() => process.env.API_POST_BODY = body)

      it('should return `axios` params from env vars', () => {
        expect(params).to.eql(axiosParams)
      })
    })

    context('when `API_QUERY_STRING` is `null`', () => {
      const queryString = process.env.API_QUERY_STRING

      let axiosParams = {}
      let params

      before(() => {
        process.env.API_QUERY_STRING = ''

        axiosParams = {
          url     : process.env.API_URL,
          method  : process.env.API_METHOD,
          headers : JSON.parse(process.env.API_HEADERS),
          data    : JSON.parse(process.env.API_POST_BODY),
          params  : null
        }

        params = StatusService.DEFAULT_AXIOS_PARAMS
      })

      after(() => process.env.API_QUERY_STRING = queryString)

      it('should return `axios` params from env vars', () => {
        expect(params).to.eql(axiosParams)
      })
    })
  })

  describe('.getAPIResponse', () => {
    context('when `params` is not `null`', () => {
      context('when `url` is not `null`', () => {
        let axiosParams = {}
        let spies = {}

        before(async () => {
          axiosParams = {
            url     : 'https://sistema.contartec.com.br/',
            method  : 'POST',
            headers : JSON.parse(process.env.API_HEADERS),
            data    : JSON.parse('{"username":"la","password":"ha"}'),
            params  : JSON.parse('{"mac":"c4:2a:fe"}')
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse(axiosParams)
        })

        after(() => SpyMock.restoreAll())

        it('should call `axios` with params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })

      context('when `url` is `null`', () => {
        let axiosParams = {}
        let spies = {}

        before(async () => {
          axiosParams = {
            url     : '',
            method  : 'POST',
            headers : JSON.parse(process.env.API_HEADERS),
            data    : JSON.parse('{"username":"la","password":"ha"}'),
            params  : JSON.parse('{"mac":"c4:2a:fe"}')
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse(axiosParams)
        })

        after(() => SpyMock.restoreAll())

        it('should not call `axios`', () => {
          expect(spies.axios).not.to.be.calledOnce
        })
      })

      context('when `headers` are `null`', () => {
        let axiosParams = {}
        let spies = {}

        before(async () => {
          axiosParams = {
            url     : 'https://sistema.contartec.com.br/',
            method  : 'POST',
            headers : null,
            data    : JSON.parse('{"username":"la","password":"ha"}'),
            params  : JSON.parse('{"mac":"c4:2a:fe"}')
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse(axiosParams)
        })

        after(() => SpyMock.restoreAll())

        it('should call `axios` with env var params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })

      context('when `data` is `null`', () => {
        let axiosParams = {}
        let spies = {}

        before(async () => {
          axiosParams = {
            url     : 'https://sistema.contartec.com.br/',
            method  : 'POST',
            headers : JSON.parse(process.env.API_HEADERS),
            data    : null,
            params  : JSON.parse('{"mac":"c4:2a:fe"}')
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse(axiosParams)
        })

        after(() => SpyMock.restoreAll())

        it('should call `axios` with env var params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })

      context('when `params` is `null`', () => {
        let axiosParams = {}
        let spies = {}

        before(async () => {
          axiosParams = {
            url     : 'https://sistema.contartec.com.br/',
            method  : 'POST',
            headers : JSON.parse(process.env.API_HEADERS),
            data    : JSON.parse('{"username":"la","password":"ha"}'),
            params  : null
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse(axiosParams)
        })

        after(() => SpyMock.restoreAll())

        it('should call `axios` with env var params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })
    })

    context('when `params` are `null`', () => {
      context('when `API_URL` is not `null`', () => {
        let axiosParams = {}
        let spies = {}

        before(async () => {
          axiosParams = {
            url     : process.env.API_URL,
            method  : process.env.API_METHOD,
            headers : JSON.parse(process.env.API_HEADERS),
            data    : JSON.parse(process.env.API_POST_BODY),
            params  : JSON.parse(process.env.API_QUERY_STRING)
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse()
        })

        after(() => SpyMock.restoreAll())

        it('should call `axios` with env var params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })

      context('when `API_URL` is `null`', () => {
        const url = process.env.API_URL

        let spies = {}

        before(async () => {
          process.env.API_URL = ''

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse()
        })

        after(async () => {
          SpyMock.restoreAll()

          process.env.API_URL = url
        })

        it('should not call `axios`', () => {
          expect(spies.axios).not.to.be.calledOnce
        })
      })

      context('when `API_HEADERS` is `null`', () => {
        const headers = process.env.API_HEADERS

        let axiosParams = {}
        let spies = {}

        before(async () => {
          process.env.API_HEADERS = ''

          axiosParams = {
            url     : process.env.API_URL,
            method  : process.env.API_METHOD,
            headers : null,
            data    : JSON.parse(process.env.API_POST_BODY),
            params  : JSON.parse(process.env.API_QUERY_STRING)
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse()
        })

        after(async () => {
          SpyMock.restoreAll()

          process.env.API_HEADERS = headers
        })

        it('should call `axios` with env var params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })

      context('when `API_POST_BODY` is `null`', () => {
        const body = process.env.API_POST_BODY

        let axiosParams = {}
        let spies = {}

        before(async () => {
          process.env.API_POST_BODY = ''

          axiosParams = {
            url     : process.env.API_URL,
            method  : process.env.API_METHOD,
            headers : JSON.parse(process.env.API_HEADERS),
            data    : null,
            params  : JSON.parse(process.env.API_QUERY_STRING)
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse()
        })

        after(async () => {
          SpyMock.restoreAll()

          process.env.API_POST_BODY = body
        })

        it('should call `axios` with env var params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })

      context('when `API_QUERY_STRING` is `null`', () => {
        const queryString = process.env.API_QUERY_STRING

        let axiosParams = {}
        let spies = {}

        before(async () => {
          process.env.API_QUERY_STRING = ''

          axiosParams = {
            url     : process.env.API_URL,
            method  : process.env.API_METHOD,
            headers : JSON.parse(process.env.API_HEADERS),
            data    : JSON.parse(process.env.API_POST_BODY),
            params  : null
          }

          spies = {
            axios: SpyMock
              .addDependencySpy(StatusService, 'axios')
          }

          await StatusService.getAPIResponse()
        })

        after(async () => {
          SpyMock.restoreAll()

          process.env.API_QUERY_STRING = queryString
        })

        it('should call `axios` with env var params', () => {
          expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
        })
      })
    })
  })

  describe('.getAPIStatus', () => {
    context('when `status: 200`', () => {
      const STATUS_CODE = 200
      const STATUS = StatusService.STATUS.OPERATIONAL

      let status

      before(async () => {
        status = await StatusService.getAPIStatus(STATUS_CODE)
      })

      it(`should return \`${STATUS}\``, () => {
        expect(status).to.eql(STATUS)
      })
    })

    context('when `status: 400`', () => {
      const STATUS_CODE = 400
      const STATUS = StatusService.STATUS.PARTIAL_OUTAGE

      let status

      before(async () => {
        status = await StatusService.getAPIStatus(STATUS_CODE)
      })

      it(`should return \`${STATUS}\``, () => {
        expect(status).to.eql(STATUS)
      })
    })

    context('when `status: 500`', () => {
      const STATUS_CODE = 500
      const STATUS = StatusService.STATUS.MAJOR_OUTAGE

      let status

      before(async () => {
        status = await StatusService.getAPIStatus(STATUS_CODE)
      })

      it(`should return \`${STATUS}\``, () => {
        expect(status).to.eql(STATUS)
      })
    })

    context('when `status` is not mapped', () => {
      const STATUS_CODE = 404
      const STATUS = StatusService.STATUS.DEGRADED_PERFORMANCE

      let status

      before(async () => {
        status = await StatusService.getAPIStatus(STATUS_CODE)
      })

      it(`should return \`${STATUS}\``, () => {
        expect(status).to.eql(STATUS)
      })
    })
  })

  describe('.updateStatusPage', () => {
    context('when `status: 1`', () => {
      const STATUS = StatusService.STATUS.OPERATIONAL
      const STATUS_CODE = 1

      let spies = {}

      before(async () => {
        const lambdaResponseMock = {
          Payload: {}
        }

        spies = {
          invoke: SpyMock
            .addDependencySpy(StatusService, 'lambda.invoke', { promise: async () => lambdaResponseMock }, false)
        }

        await StatusService.updateStatusPage(STATUS)
      })

      after(() => SpyMock.restoreAll())

      it('should update the `component` with `status: 1', () => {
        const payload = spies.invoke
          .getCall(0).args[0].Payload

        expect(JSON.parse(payload).body.status).to.eql(STATUS_CODE)
      })
    })

    context('when `status: 4`', () => {
      const STATUS = StatusService.STATUS.PARTIAL_OUTAGE
      const STATUS_CODE = 4

      let spies = {}

      before(async () => {
        const lambdaResponseMock = {
          Payload: {}
        }

        spies = {
          invoke: SpyMock
            .addDependencySpy(StatusService, 'lambda.invoke', { promise: async () => lambdaResponseMock }, false)
        }

        await StatusService.updateStatusPage(STATUS)
      })

      after(() => SpyMock.restoreAll())

      it('should update the `component` with `status: 4', () => {
        const payload = spies.invoke
          .getCall(0).args[0].Payload

        expect(JSON.parse(payload).body.status).to.eql(STATUS_CODE)
      })
    })

    context('when `componentIds` is not `null`', () => {
      const STATUS = StatusService.STATUS.OPERATIONAL
      const COMPONENT_IDS = '1,2,3'

      let spies = {}

      before(async () => {
        const lambdaResponseMock = {
          Payload: {}
        }

        spies = {
          invoke: SpyMock
            .addDependencySpy(StatusService, 'lambda.invoke', { promise: async () => lambdaResponseMock }, false)
        }

        await StatusService.updateStatusPage(STATUS, COMPONENT_IDS)
      })

      after(() => SpyMock.restoreAll())

      it('should update the `componentIds` from params', () => {
        const payload = spies.invoke
          .getCall(0).args[0].Payload

        expect(JSON.parse(payload).body.componentIds).to.eql(COMPONENT_IDS)
      })
    })

    context('when `componentIds` is `null`', () => {
      const STATUS = StatusService.STATUS.OPERATIONAL
      const COMPONENT_IDS = process.env.COMPONENT_IDS

      let spies = {}

      before(async () => {
        const lambdaResponseMock = {
          Payload: {}
        }

        spies = {
          invoke: SpyMock
            .addDependencySpy(StatusService, 'lambda.invoke', { promise: async () => lambdaResponseMock }, false)
        }

        await StatusService.updateStatusPage(STATUS)
      })

      after(() => SpyMock.restoreAll())

      it('should update the `componentIds` from env vars', () => {
        const payload = spies.invoke
          .getCall(0).args[0].Payload

        expect(JSON.parse(payload).body.componentIds).to.eql(COMPONENT_IDS)
      })
    })
  })

  describe('.updateStatus', () => {
    context('when `webAPIResponse` is not `null`', () => {
      context('when `webAPIResponse` is `200`', () => {
        const STATUS = 200
        const STATUS_CODE = StatusService.STATUS.OPERATIONAL

        let response = null
        let spies = {}

        before(async () => {
          spies = {
            getAPIResponse: SpyMock
              .addReturnSpy(StatusService, 'getAPIResponse', { status : STATUS }),

            updateStatusPage: SpyMock
              .addReturnSpy(StatusService, 'updateStatusPage')
          }

          response = await StatusService.updateStatus()
        })

        after(() => SpyMock.restoreAll())

        it('should return the `API` `statusCode`', () => {
          expect(response).to.eql(STATUS_CODE)
        })

        it('should call `.updateStatusPage` once', () => {
          expect(spies.updateStatusPage).to.be.calledOnce
        })
      })

      context('when `webAPIResponse` is `400 || 500`', () => {
        const STATUS = 400
        const STATUS_CODE = StatusService.STATUS.PARTIAL_OUTAGE

        let response = null
        let spies = {}

        before(async () => {
          const apiMock = {
            response: { status : STATUS }
          }

          spies = {
            getAPIResponse: SpyMock
              .addExceptionSpy(StatusService, 'getAPIResponse', apiMock),

            updateStatusPage: SpyMock
              .addReturnSpy(StatusService, 'updateStatusPage')
          }

          response = await StatusService.updateStatus()
        })

        after(() => SpyMock.restoreAll())

        it('should return the `API` `statusCode`', () => {
          expect(response).to.eql(STATUS_CODE)
        })

        it('should call `.updateStatusPage` once', () => {
          expect(spies.updateStatusPage).to.be.calledOnce
        })
      })
    })

    context('when `webAPIResponse`is `null`', () => {
      let status = null

      before(async () => {
        SpyMock
          .addReturnSpy(StatusService, 'getAPIResponse')

        status = await StatusService.updateStatus()
      })

      after(() => SpyMock.restoreAll())

      it('should return `null`', () => {
        expect(status).not.to.exist
      })
    })
  })
})

tests/services/StatusPageIOService.test.js
'use strict'

require('dotenv').config()

const rewire = require('rewire')

const StatusPageIOService = rewire('../../lib/services/StatusPageIOService')

const SpyMock = require('@contartec-team/spy-mock/lib/SpyMock')

describe('StatusPageIOService', () => {
  describe('updateComponentStatus', () => {
    context('when `id` and `status` are not `null`', () => {
      const ID = 'la'
      const status = 'operational'

      let axiosParams = {}
      let spies = {}

      before(async () => {
        axiosParams = { 
          url     : 'https://api.statuspage.io/v1/pages/lalaId/components/la',
          method  : 'PUT',
          headers : StatusPageIOService.DEFAULT_HEADERS,
          data    : {
            component: {
              status
            }
          }
        }

        spies = {
          axios: SpyMock
            .addDependencySpy(StatusPageIOService, 'axios')
        }

        await StatusPageIOService.updateComponentStatus(ID, status)
      })

      after(() => SpyMock.restoreAll())

      it('should update the `components`', () => {
        expect(spies.axios.getCall(0).args[0]).to.eql(axiosParams)
      })
    })

    context('when `id` and `status` are `null`', () => {
      let spies = {}

      before(async () => {
        spies = {
          axios: SpyMock
            .addDependencySpy(StatusPageIOService, 'axios')
        }

        await StatusPageIOService.updateComponentStatus()
      })

      after(() => SpyMock.restoreAll())

      it('should not handle the `event`', () => {
        expect(spies.axios).not.to.be.calledOnce
      })
    })
  })

  describe('updateComponentsStatus', () => {
    context('when `componentIds` and `status` are not `null`', () => {
      const componentIds = ['la', 'le', 'ha']
      const status = 'operational'

      let spies = {}

      before(async () => {
        spies = {
          updateComponentStatus: SpyMock
            .addReturnSpy(StatusPageIOService, 'updateComponentStatus')
        }

        await StatusPageIOService.updateComponentsStatus(componentIds, status)
      })

      after(() => SpyMock.restoreAll())

      it('should update the `components`', () => {
        expect(spies.updateComponentStatus).to.be.calledThrice
      })
    })

    context('when `componentIds` and `status` are `null`', () => {
      let spies = {}

      before(async () => {
        spies = {
          updateComponentStatus: SpyMock
            .addReturnSpy(StatusPageIOService, 'updateComponentStatus')
        }

        await StatusPageIOService.updateComponentsStatus()
      })

      after(() => SpyMock.restoreAll())

      it('should update the `components`', () => {
        expect(spies.updateComponentStatus).not.to.be.calledOnce
      })
    })
  })
})

lib/services/StatusPageIOService.js
'use strict'

const axios = require('axios')

const API_KEY = process.env.STATUS_PAGE_API_KEY
const DEFAULT_PAGE_ID = process.env.STATUS_PAGE_PAGE_ID

const URL = 'https://api.statuspage.io/v1/pages'

/**
 * Service that handle updates to [`Status Page`](https://www.atlassian.com/software/statuspage)
 * @class StatusPageIOService
*/
class StatusPageIOService {

  /**
   * The list of `component` `status`
   * @enum {Object} COMPONENT_STATUS
  */
  static get STATUS() {
    return {
      OPERATIONAL           : 'operational',
      UNDER_MAINTENANCE     : 'under_maintenance',
      DEGRADED_PERFORMANCE  : 'degraded_performance',
      PARTIAL_OUTAGE        : 'partial_outage',
      MAJOR_OUTAGE          : 'major_outage'
    }
  }

  /**
   * The list of `component` `status` numbers
   * @enum {Object} COMPONENT_STATUS_NUMBER
  */
  static get STATUS_NUMBER() {
    return {
      1: StatusPageIOService.STATUS.OPERATIONAL,
      2: StatusPageIOService.STATUS.UNDER_MAINTENANCE,
      3: StatusPageIOService.STATUS.DEGRADED_PERFORMANCE,
      4: StatusPageIOService.STATUS.PARTIAL_OUTAGE,
      5: StatusPageIOService.STATUS.MAJOR_OUTAGE
    }
  }

  /**
   * Default `headers`
   * @type {Object}
  */
  static get DEFAULT_HEADERS() {
    return {
      Authorization: `OAuth ${API_KEY}`
    }
  }

  /**
   * Updates the `component` status
   *
   * @param {string} id The `component` id
   * @param {string} status The `component` status (@see {link StatusPageIOService.COMPONENT_STATUS})
   * @param {string} [pageId] The `page` id
   *
   * @return {Promise<Object>} The http response
  */
  static updateComponentStatus(id, status, pageId = DEFAULT_PAGE_ID) {
    let promise = null

    if (id && status) {
      const url = `${URL}/${pageId}/components/${id}`

      promise = axios({
        url,
        method  : 'PUT',
        headers : StatusPageIOService.DEFAULT_HEADERS,
        data    : {
          component: {
            status
          }
        }
      })
    }

    return promise
  }

  /**
   * Updates the `components` status
   *
   * @param {string | Array<string>} componentIds The `components` ids
   * @param {string} status The `component` status (@see {link StatusPageIOService.COMPONENT_STATUS})
   * @param {string} [pageId] The `page` id
   *
   * @return {Promise<Array>} The http responses
  */
  static updateComponentsStatus(componentIds, status, pageId = DEFAULT_PAGE_ID) {
    let responses = []

    if (componentIds && status) {
      let componentIdsTemp = [ ...componentIds ]

      if (typeof(componentIds) === 'string') {
        componentIdsTemp = componentIds
          .replace(/ /g, '')
          .split(',')
      }

      const promises = componentIdsTemp
        .map(id => {
          return StatusPageIOService
            .updateComponentStatus(id, status, pageId)
        })
      
      responses = Promise.all(promises)
    }

    return responses
  }
}

module.exports = StatusPageIOService

lib/services/StatusService.js
'use strict'

const AWS = require('aws-sdk')
const axios = require('axios')

const lambda = new AWS.Lambda({
  region: 'us-west-2'
})

const DEFAULT_LAMBDA_PARAMS = {
  FunctionName    : `status-page-update-${process.env.STAGE}-http`,
  InvocationType  : 'RequestResponse'
}

/**
 * Status page status list
 * @type {Object}
 * @memberof StatusService
*/
const STATUS = {
  OPERATIONAL           : 1,
  UNDER_MAINTENANCE     : 2,
  DEGRADED_PERFORMANCE  : 3,
  PARTIAL_OUTAGE        : 4,
  MAJOR_OUTAGE          : 5
}

const COMPONENT_IDS = process.env.COMPONENT_IDS

/**
 * Handle `Status Page` updates
 * @class StatusService
*/
class StatusService {

  static get STATUS()           { return STATUS }

  static get API_URL()          { return process.env.API_URL }
  static get API_METHOD()       { return process.env.API_METHOD }
  static get API_QUERY_STRING() { return process.env.API_QUERY_STRING }
  static get API_POST_BODY()    { return process.env.API_POST_BODY }
  static get API_HEADERS()      { return process.env.API_HEADERS }


  /**
   * Return the env var params
   *
   * @return {Object} The API env vars
  */
  static get DEFAULT_AXIOS_PARAMS() {
    const data = StatusService.API_POST_BODY ?
      JSON.parse(StatusService.API_POST_BODY) :
      null

    const params = StatusService.API_QUERY_STRING ?
      JSON.parse(StatusService.API_QUERY_STRING) :
      null

    const headers = StatusService.API_HEADERS ?
      JSON.parse(StatusService.API_HEADERS) :
      null

    const defaultParams = {
      url     : StatusService.API_URL,
      method  : StatusService.API_METHOD,
      headers,
      data,
      params
    }

    return defaultParams
  }

  /**
   * Return the web API
   *
   * @param {Object} params The axios params
   * @param {Object} params.url The server url
   * @param {Object} params.method The request method
   * @param {Object} params.headers The headers request
   * @param {Object} params.data The body request
   * @param {Object} params.params The URL parameters
   *
   * @return {Object} The axios response
  */
  static getAPIResponse(params = {}) {
    let webAPIResponse = null

    const paramsTemp = {
      ...StatusService.DEFAULT_AXIOS_PARAMS,
      ...params
    }

    if (paramsTemp.url)
      webAPIResponse = axios(paramsTemp)

    return webAPIResponse
  }

  /**
   * Return the API status
   *
   * @param {(StatusService.STATUS | Number)} status The http status code
   *
   * @return {StatusService.STATUS} The API status
  */
  static getAPIStatus(status) {
    let response = null

    switch (status) {
      case 200:
        response = STATUS.OPERATIONAL
        break
      case 400:
        response = STATUS.PARTIAL_OUTAGE
        break
      case 500:
        response = STATUS.MAJOR_OUTAGE
        break
      default:
        response = STATUS.DEGRADED_PERFORMANCE
        break
    }

    return response
  }

  /**
   * Updates the `Status Page` with `status`
   *
   * @param {(StatusService.STATUS | Number)} status The new status
   * @param {Array<string>} [componentIds = COMPONENT_IDS] The `Status Page` components ids separeted by comma
   *
   * @return {Promise} The lambda response invocation
  */
  static updateStatusPage(status, componentIds = COMPONENT_IDS) {
    const lambdaParams = {
      ...DEFAULT_LAMBDA_PARAMS,
      Payload: JSON
        .stringify({
          body: {
            status,
            componentIds
          }
        })
    }

    return lambda
      .invoke(lambdaParams)
      .promise()
  }

  /**
   * Updates the component(s) `status`
   * @async
   *
   * @param {Object} [params] The axios params (@see {@link StatusService.getAPIResponse})
   *
   * @return {StatusService.STATUS} The API status
  */
  static async updateStatus(params = {}) {
    let webAPIResponse = null

    try {
      webAPIResponse = await StatusService
        .getAPIResponse(params)

      console
        .warn(`[StatusService.updateStatus] ${new Date().toJSON()} webAPIResponse`, {
          webAPIResponse
        })
    }
    catch (e) {
      webAPIResponse = e.response

      console
        .error(`[StatusService.updateStatus] ${new Date().toJSON()} error`, {
          error: e.message,
          stack: e.stack,
          e
        })
    }

    let status = null

    if (webAPIResponse) {
      console
        .warn(`[StatusService.updateStatus] ${new Date().toJSON()} webAPIResponse ${webAPIResponse.status}`, {
          body    : webAPIResponse.body,
          headers : webAPIResponse.headers
        })

      status = StatusService.getAPIStatus(webAPIResponse.status)

      await StatusService.updateStatusPage(status)
    }

    return status
  }
}

module.exports = StatusService